File: frontend/lib/geography.ts
'use client';

import { useEffect, useState } from 'react';
import { api } from '@/lib/api';

export const REGION_OPTIONS = [
    'Central Asia',
    'Asia',
    'Europe',
    'Africa',
    'North America',
    'Latin America',
] as const;

export const CENTRAL_ASIA_REGION = 'Central Asia';

export const CENTRAL_ASIA_COUNTRIES = [
    'Uzbekistan',
    'Kazakhstan',
    'Kyrgyzstan',
    'Tajikistan',
    'Turkmenistan',
] as const;

export type GeographyMeta = {
    regions: string[];
    countries_by_region: Record<string, string[]>;
    central_asia_countries: string[];
};

export const DEFAULT_GEOGRAPHY_META: GeographyMeta = {
    regions: [...REGION_OPTIONS],
    countries_by_region: {
        [CENTRAL_ASIA_REGION]: [...CENTRAL_ASIA_COUNTRIES],
        Asia: [],
        Europe: [],
        Africa: [],
        'North America': [],
        'Latin America': [],
    },
    central_asia_countries: [...CENTRAL_ASIA_COUNTRIES],
};

function normalizeList(values: unknown, fallback: string[]): string[] {
    if (!Array.isArray(values)) {
        return fallback;
    }

    const normalized: string[] = [];
    const seen = new Set<string>();

    values.forEach((value) => {
        if (typeof value !== 'string') {
            return;
        }
        const cleaned = value.trim();
        const key = cleaned.toLowerCase();
        if (cleaned && !seen.has(key)) {
            normalized.push(cleaned);
            seen.add(key);
        }
    });

    return normalized.length > 0 ? normalized : fallback;
}

function normalizeCountriesByRegion(
    value: unknown,
    regions: string[],
): Record<string, string[]> {
    const fallback = DEFAULT_GEOGRAPHY_META.countries_by_region;
    if (!value || typeof value !== 'object' || Array.isArray(value)) {
        return fallback;
    }

    return regions.reduce<Record<string, string[]>>((current, region) => {
        current[region] = normalizeList(
            (value as Record<string, unknown>)[region],
            fallback[region] ?? [],
        );
        return current;
    }, {});
}

export function normalizeGeographyMeta(data: Partial<GeographyMeta> | null | undefined): GeographyMeta {
    const regions = normalizeList(data?.regions, DEFAULT_GEOGRAPHY_META.regions);
    const centralAsiaCountries = normalizeList(
        data?.central_asia_countries,
        DEFAULT_GEOGRAPHY_META.central_asia_countries,
    );
    const countriesByRegion = normalizeCountriesByRegion(data?.countries_by_region, regions);

    countriesByRegion[CENTRAL_ASIA_REGION] = centralAsiaCountries;

    return {
        regions,
        countries_by_region: countriesByRegion,
        central_asia_countries: centralAsiaCountries,
    };
}

export function useGeographyMeta(): GeographyMeta {
    const [geography, setGeography] = useState<GeographyMeta>(DEFAULT_GEOGRAPHY_META);

    useEffect(() => {
        let mounted = true;

        api.get<Partial<GeographyMeta>>('/meta/geography')
            .then((response) => {
                if (mounted) {
                    setGeography(normalizeGeographyMeta(response.data));
                }
            })
            .catch(() => {
                if (mounted) {
                    setGeography(DEFAULT_GEOGRAPHY_META);
                }
            });

        return () => {
            mounted = false;
        };
    }, []);

    return geography;
}

File: backend/app/core/geography.py
from __future__ import annotations

from collections.abc import Iterable


REGION_OPTIONS: tuple[str, ...] = (
    "Central Asia",
    "Asia",
    "Europe",
    "Africa",
    "North America",
    "Latin America",
)

CENTRAL_ASIA_REGION = "Central Asia"

CENTRAL_ASIA_COUNTRIES: tuple[str, ...] = (
    "Uzbekistan",
    "Kazakhstan",
    "Kyrgyzstan",
    "Tajikistan",
    "Turkmenistan",
)

COUNTRIES_BY_REGION: dict[str, tuple[str, ...]] = {
    CENTRAL_ASIA_REGION: CENTRAL_ASIA_COUNTRIES,
    "Asia": (),
    "Europe": (),
    "Africa": (),
    "North America": (),
    "Latin America": (),
}


def _canonical_lookup(values: Iterable[str]) -> dict[str, str]:
    return {value.casefold(): value for value in values}


def normalize_allowed_values(
    values: Iterable[str],
    allowed_values: Iterable[str],
    *,
    label: str,
    reject_invalid: bool = True,
) -> list[str]:
    """Trim, dedupe, and canonicalize values against a known option list."""
    lookup = _canonical_lookup(allowed_values)
    normalized: list[str] = []
    seen: set[str] = set()

    for item in values:
        if not isinstance(item, str):
            if reject_invalid:
                raise ValueError(f"Invalid {label}: {item}")
            continue

        cleaned = item.strip()
        if not cleaned:
            continue

        canonical = lookup.get(cleaned.casefold())
        if canonical is None:
            if reject_invalid:
                raise ValueError(f"Invalid {label}: {cleaned}")
            continue

        key = canonical.casefold()
        if key not in seen:
            normalized.append(canonical)
            seen.add(key)

    return normalized


def normalize_target_regions(
    values: Iterable[str] | None,
    *,
    reject_invalid: bool = True,
) -> list[str] | None:
    if values is None:
        return None
    return normalize_allowed_values(
        values,
        REGION_OPTIONS,
        label="target region",
        reject_invalid=reject_invalid,
    )


def normalize_target_countries(
    values: Iterable[str] | None,
    *,
    reject_invalid: bool = True,
) -> list[str] | None:
    if values is None:
        return None
    return normalize_allowed_values(
        values,
        CENTRAL_ASIA_COUNTRIES,
        label="target country",
        reject_invalid=reject_invalid,
    )


def geography_meta_payload() -> dict[str, object]:
    return {
        "regions": list(REGION_OPTIONS),
        "countries_by_region": {
            region: list(COUNTRIES_BY_REGION.get(region, ()))
            for region in REGION_OPTIONS
        },
        "central_asia_countries": list(CENTRAL_ASIA_COUNTRIES),
    }
